opm/simulators/flow/BlackoilModelEbos.hpp

```diff
@@ -472,13 +625,346 @@ namespace Opm {
             return report;
         }
 
+
+
+
+        template <class NonlinearSolverType>
+        SimulatorReportSingle nonlinearIterationNldd(const int iteration,
+                                                     const SimulatorTimerInterface& timer,
+                                                     NonlinearSolverType& nonlinear_solver)
+        {
+            // -----------   Set up reports and timer   -----------
+            SimulatorReportSingle report;
+            failureReport_ = SimulatorReportSingle();
+            Dune::Timer perfTimer;
+
+            perfTimer.start();
+            report.total_linearizations = 1;
+
+            // -----------   Assemble   -----------
+            try {
+                report += assembleReservoir(timer, iteration);
+                report.assemble_time += perfTimer.stop();
+            }
+            catch (...) {
+                report.assemble_time += perfTimer.stop();
+                failureReport_ += report;
+                throw; // continue throwing the stick
+            }
+
+            // -----------   Check if converged   -----------
+            std::vector<double> residual_norms;
+            perfTimer.reset();
+            perfTimer.start();
+            // the step is not considered converged until at least minIter iterations is done
+            {
+                auto convrep = getConvergence(timer, iteration, residual_norms);
+                report.converged = convrep.converged()  && iteration > nonlinear_solver.minIter();;
+                ConvergenceReport::Severity severity = convrep.severityOfWorstFailure();
+                convergence_reports_.back().report.push_back(std::move(convrep));
+
+                // Throw if any NaN or too large residual found.
+                if (severity == ConvergenceReport::Severity::NotANumber) {
+                    OPM_THROW(NumericalProblem, "NaN residual found!");
+                } else if (severity == ConvergenceReport::Severity::TooLarge) {
+                    OPM_THROW_NOLOG(NumericalProblem, "Too large residual found!");
+                }
+            }
+            report.update_time += perfTimer.stop();
+            residual_norms_history_.push_back(residual_norms);
+
+            if (report.converged) {
+                return report;
+            }
+
+            // -----------   If not converged, do an NLDD iteration   -----------
+
+            auto& solution = ebosSimulator().model().solution(0);
+            auto initial_solution = solution;
+            auto locally_solved = initial_solution;
+
+            // -----------   Decide on an ordering for the domains   -----------
+            std::vector<int> domain_order(domains_.size());
+            if (param_.local_solve_approach_ == "gauss-seidel") {
+                // TODO: enable flexibility and choice in choosing domain ordering approach.
+                if (true) {
```

Comment:
Will this ever be anything other than `true`?   If not, we can probably save a level of nesting.

Reply:
I would like to keep the extra code with the TODO above as a reminder to look into this, if that is ok.